packages/babel-plugin-proposal-private-methods/src/index.js

```diff
@@ -3,20 +3,255 @@
 import { declare } from "@babel/helper-plugin-utils";
 import {
   createClassFeaturePlugin,
+  enableFeatureIfCompilingFields,
   FEATURES,
 } from "@babel/helper-create-class-features-plugin";
+import ReplaceSupers, {
+  environmentVisitor,
+} from "@babel/helper-replace-supers";
+import { isRequired } from "@babel/helper-compilation-targets";
 
 export default declare((api, options) => {
   api.assertVersion(7);
 
-  return createClassFeaturePlugin({
-    name: "proposal-private-methods",
+  if (!api.targets || isRequired("proposal-class-properties", api.targets())) {
+    return createClassFeaturePlugin({
+      name: "proposal-private-methods",
+
+      feature: FEATURES.privateMethods,
+      loose: options.loose,
+
+      manipulateOptions(opts, parserOpts) {
+        parserOpts.plugins.push("classPrivateMethods");
+      },
+    });
+  }
+
+  const { types: t, template, traverse } = api;
+
+  function readOnlyError(member, file) {
+    return t.callExpression(file.addHelper("readOnlyError"), [
+      t.stringLiteral(`#${member.property.id.name}`),
+    ]);
+  }
+
+  function isKnownPrivate(node, readonlyNames) {
+    return (
+      t.isMemberExpression(node) &&
+      t.isPrivateName(node.property) &&
+      readonlyNames.has(node.property.id.name)
+    );
+  }
 
-    feature: FEATURES.privateMethods,
-    loose: options.loose,
+  function toMemoizedFunction(node, scope, hint) {
+    if (node.static) return t.toExpression(node);
+
+    const tmpId = scope.generateUidIdentifier(hint);
+    scope.push({ id: tmpId });
+
+    return template.expression.ast`
+        ${t.cloneNode(tmpId)} || (
+          ${t.cloneNode(tmpId)} = ${t.toExpression(node)}
+        )
+      `;
+  }
+
+  const privateUsageVisitor = {
+    AssignmentExpression(path, { readonly, file }) {
+      const { left, right } = path.node;
+      if (isKnownPrivate(left, readonly)) {
+        const seq = [];
+
+        if (!path.get("left.object").isPure()) seq.push(left.object);
+        if (!path.get("right").isPure()) seq.push(right);
+        seq.push(readOnlyError(left, file));
+
+        path.replaceWith(t.sequenceExpression(seq));
+      }
+    },
+    UpdateExpression(path, { readonly, file }) {
+      const { argument } = path.node;
+      if (isKnownPrivate(argument, readonly)) {
+        const seq = [];
+
+        if (!path.get("argument.object").isPure()) seq.push(argument.object);
+        seq.push(readOnlyError(argument, file));
+
+        path.replaceWith(t.sequenceExpression(seq));
+      }
+    },
+    MemberExpression: {
+      exit(path, { accessors }) {
+        if (isKnownPrivate(path.node, accessors)) {
+          path.replaceWith(t.memberExpression(path.node, t.identifier("_")));
+          path.skip();
+        }
+      },
+    },
+
+    // Don't transform usage of shadowed private names
+    ClassBody(path, oldState) {
```

Comment:
Consider reuse `environmentVisitor` from packages/babel-helper-replace-supers/src/index.js, since computed keys are meant to be resolved in upper private scope.

```js
class C {
  get #foo() { return 42 }
  constructor () {
    class S {
      [this.#foo]
  }
}
```